### ISSUE TITLE
Superfluous ShadowingOuterLocalVariable in if-condition

### ISSUE THREAD

**Issue by sudoremo** [rubocop/rubocop#11834]
Consider the following example:

```ruby
class Sample < RailsOps::Operation
  def my_method
    if (rand * 100).to_i.even?
      id = 1
      @something = id
    else
      ids = [1, 2, 3]

      ids.each do |id|
        @something = id
      end
    end
  end
end

```

--------

## Expected behavior

Rubocop should not show any offenses.

## Actual behavior

Rubocop shows the following offense:

```
rubocop --force-default-config --debug sample.rb

Default configuration from /my/path/rubocop-1.42.0/config/default.yml
Inspecting 1 file
Scanning /my/path/sample.rb
W

Offenses:

sample.rb:12:20: W: Lint/ShadowingOuterLocalVariable: Shadowing outer local variable - id.
      ids.each do |id|
                   ^^

1 file inspected, 1 offense detected
Finished in 0.2472049999050796 seconds

```

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
rubocop -V

1.42.0 (using Parser 3.2.1.0, rubocop-ast 1.24.1, running on ruby 3.2.1) [x86_64-darwin22]
  - rubocop-rails 2.15.2
```

**Comment by jonas054** [comment]
@sudoremo I don't think there's a bug here. Look at this variation on your code:

```ruby
def my_method(condition)
  if condition
    id = 1
  else
    ids = [2, 3, 4]
    ids.each do |id2|
      id = 7
      p [id, id2]
    end
  end
  p condition => id
end

my_method(true)
puts
my_method(false)
```

It outputs

```ruby
{true=>1}

[7, 2]
[7, 3]
[7, 4]
{false=>7}
```

By choosing a different name (`id2`) for the block parameter, we can still access the outer variable `id` inside the block. It's defined and has the value `nil` befor we set it to `7`. If we name the block parameter `id`, we can't access the outer variable. It's shadowed.

**Comment by alexeyschepin** [comment]
@jonas054, I believe there is still a bug. I was able to replicate it with the following test in the latest rubocop version

```ruby
context 'when a block local variable has same name as an outer scope variable' \
        'with different branches of same `if` condition node' \
        'and located in a begin node' do
  it 'does not register an offense' do
    expect_no_offenses(<<~RUBY)
      def some_method
        if condition?
          foo = 1
        else
          bar = [1, 2, 3]
          bar.each do |foo|
          end
        end
      end
    RUBY
  end
end
```
This test fails unless you delete `bar = [1, 2, 3]` line. The problem is not in the content of this line but the fact that `foo` variable has an extra parent node. Suddenly `variable_node` becomes `begin` instead of `if` on line https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/lint/shadowing_outer_local_variable.rb#L71.
So `same_conditions_node_different_branch?` returns `false`.

I am trying to understand how exactly everything works here before proposing any solution.

**Comment by jonas054** [comment]
Thanks @alexeyschepin! I agree. That's a bug.

**Comment by sudoremo** [comment]
Many thanks for the fix 👍 Do you have any idea when this might be released?
